Mathlib/MeasureTheory/Measure/LogLikelihoodRatio.lean

```diff
@@ -0,0 +1,200 @@
+lemma exp_llr_of_ac' (μ ν : Measure α) [SigmaFinite μ] [SigmaFinite ν] (hμν : ν ≪ μ) :
+    (fun x ↦ exp (llr μ ν x)) =ᵐ[ν] fun x ↦ (μ.rnDeriv ν x).toReal := by
+  filter_upwards [exp_llr μ ν, Measure.rnDeriv_pos' hμν] with x hx hx_pos
+  rwa [if_neg hx_pos.ne'] at hx
+
+lemma neg_llr [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν) :
+    - llr μ ν =ᵐ[μ] llr ν μ := by
+  filter_upwards [Measure.inv_rnDeriv hμν] with x hx
+  rw [Pi.neg_apply, llr, llr, ← log_inv, ← ENNReal.toReal_inv]
+  congr
+
+lemma exp_neg_llr [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν) :
+    (fun x ↦ exp (- llr μ ν x)) =ᵐ[μ] fun x ↦ (ν.rnDeriv μ x).toReal := by
+  filter_upwards [neg_llr hμν, exp_llr_of_ac' ν μ hμν] with x hx hx_exp_log
+  rw [Pi.neg_apply] at hx
+  rw [hx, hx_exp_log]
+
+lemma exp_neg_llr' [SigmaFinite μ] [SigmaFinite ν] (hμν : ν ≪ μ) :
+    (fun x ↦ exp (- llr μ ν x)) =ᵐ[ν] fun x ↦ (ν.rnDeriv μ x).toReal := by
+  filter_upwards [neg_llr hμν, exp_llr_of_ac ν μ hμν] with x hx hx_exp_log
+  rw [Pi.neg_apply, neg_eq_iff_eq_neg] at hx
+  rw [← hx, hx_exp_log]
+
+@[measurability]
+lemma measurable_llr (μ ν : Measure α) : Measurable (llr μ ν) :=
+  (Measure.measurable_rnDeriv μ ν).ennreal_toReal.log
+
+@[measurability]
+lemma stronglyMeasurable_llr (μ ν : Measure α) : StronglyMeasurable (llr μ ν) :=
+  (measurable_llr μ ν).stronglyMeasurable
+
+lemma llr_smul_left [IsFiniteMeasure μ] [Measure.HaveLebesgueDecomposition μ ν]
+    (hμν : μ ≪ ν) (c : ℝ≥0∞) (hc : c ≠ 0) (hc_ne_top : c ≠ ∞) :
+    llr (c • μ) ν =ᵐ[μ] fun x ↦ llr μ ν x + log c.toReal := by
+  simp only [llr, llr_def]
+  have h := Measure.rnDeriv_smul_left_of_ne_top μ ν hc_ne_top
+  filter_upwards [hμν.ae_le h, Measure.rnDeriv_pos hμν, hμν.ae_le (Measure.rnDeriv_lt_top μ ν)]
+    with x hx_eq hx_pos hx_ne_top
+  rw [hx_eq]
+  simp only [Pi.smul_apply, smul_eq_mul, ENNReal.toReal_mul]
+  rw [log_mul]
+  rotate_left
+  · rw [ENNReal.toReal_ne_zero]
+    simp [hc, hc_ne_top]
+  · rw [ENNReal.toReal_ne_zero]
+    simp [hx_pos.ne', hx_ne_top.ne]
+  ring
+
+lemma llr_smul_right [IsFiniteMeasure μ] [Measure.HaveLebesgueDecomposition μ ν]
+    (hμν : μ ≪ ν) (c : ℝ≥0∞) (hc : c ≠ 0) (hc_ne_top : c ≠ ∞) :
+    llr μ (c • ν) =ᵐ[μ] fun x ↦ llr μ ν x - log c.toReal := by
+  simp only [llr, llr_def]
+  have h := Measure.rnDeriv_smul_right_of_ne_top μ ν hc hc_ne_top
+  filter_upwards [hμν.ae_le h, Measure.rnDeriv_pos hμν, hμν.ae_le (Measure.rnDeriv_lt_top μ ν)]
+    with x hx_eq hx_pos hx_ne_top
+  rw [hx_eq]
+  simp only [Pi.smul_apply, smul_eq_mul, ENNReal.toReal_mul]
+  rw [log_mul]
+  rotate_left
+  · rw [ENNReal.toReal_ne_zero]
+    simp [hc, hc_ne_top]
+  · rw [ENNReal.toReal_ne_zero]
+    simp [hx_pos.ne', hx_ne_top.ne]
+  rw [ENNReal.toReal_inv, log_inv]
+  ring
+
+section llr_tilted
+
+lemma llr_tilted_left [SigmaFinite μ] [SigmaFinite ν] (hμν : μ ≪ ν)
+    (hf : Integrable (fun x ↦ exp (f x)) μ) (hfν : AEMeasurable f ν) :
+    (llr (μ.tilted f) ν) =ᵐ[μ] fun x ↦ f x - log (∫ x, exp (f x) ∂μ) + llr μ ν x := by
```

Comment:
```suggestion
    (hf : Integrable (fun x ↦ exp (f x)) μ) (hfν : AEMeasurable f ν) :
    (llr (μ.tilted f) ν) =ᵐ[μ] fun x ↦ f x - log (∫ z, exp (f z) ∂μ) + llr μ ν x := by
```
I think it's nicer to avoid reusing `x` here.